### PR TITLE
[Coop] Convert Mono.Interop.ComInteropProxy to coop.

### PR DESCRIPTION
**Pull request by jaykrell**

### CODE REVIEW

File: mono/metadata/cominterop.c [hunk 1]
```diff
@@ -1871,17 +1871,16 @@ ves_icall_System_ComObject_GetInterfaceInternal (MonoComObject* obj, MonoReflect
...
 #ifndef DISABLE_COM
-	guint32 gchandle = 0;
 	if (!rcw_hash) {
 		mono_cominterop_lock ();
 		rcw_hash = g_hash_table_new (mono_aligned_addr_hash, NULL);
 		mono_cominterop_unlock ();
 	}
 
-	gchandle = mono_gchandle_new_weakref ((MonoObject*)proxy, FALSE);
+	guint32 const gchandle = mono_gchandle_new_weakref_from_handle ((MonoObjectHandle)proxy);
```
Comment (lambdageek) [inline mono/metadata/cominterop.c]: nit `MONO_HANDLE_CAST (MonoObject, proxy)`

File: mono/metadata/cominterop.c [hunk 2]
```diff
@@ -1891,24 +1890,24 @@ ves_icall_Mono_Interop_ComInteropProxy_AddProxy (gpointer pUnk, MonoComInteropPr
...
-		proxy = (MonoComInteropProxy*)mono_gchandle_get_target (gchandle);
-		/* proxy is null means we need to free up old RCW */
-		if (!proxy) {
-			mono_gchandle_free (gchandle);
-			g_hash_table_remove (rcw_hash, pUnk);
-		}
+	if (!gchandle)
+		return MONO_HANDLE_NEW (MonoComInteropProxy, NULL);
+
+	MonoComInteropProxyHandle const proxy = (MonoComInteropProxyHandle)mono_gchandle_get_target_handle (gchandle);
```
Comment (lambdageek) [inline mono/metadata/cominterop.c]: nit `MONO_HANDLE_CAST (MonoComInteropProxy, ...)`

File: mono/metadata/marshal.c [hunk 3]
```diff
@@ -364,7 +364,7 @@ mono_delegate_handle_to_ftnptr (MonoDelegateHandle delegate, MonoError *error)
 	MonoObjectHandle delegate_target = MONO_HANDLE_NEW_GET (MonoObject, delegate, target);
 	if (!MONO_HANDLE_IS_NULL (delegate_target)) {
 		/* Produce a location which can be embedded in JITted code */
-		target_handle = mono_gchandle_new_weakref (MONO_HANDLE_RAW (delegate_target), FALSE); /* FIXME: a version of mono_gchandle_new_weakref that takes a coop handle */
+		target_handle = mono_gchandle_new_weakref_from_handle ((MonoObjectHandle)delegate_target);
```
Comment (lambdageek) [inline mono/metadata/marshal.c]: redundant cast

File: mono/metadata/marshal.c [hunk 4]
```diff
@@ -426,7 +426,7 @@ delegate_hash_table_add (MonoDelegateHandle d)
 		delegate_hash_table = delegate_hash_table_new ();
 	gpointer delegate_trampoline = MONO_HANDLE_GETVAL (d, delegate_trampoline);
 	if (mono_gc_is_moving ()) {
-		gchandle = mono_gchandle_new_weakref ((MonoObject*) MONO_HANDLE_RAW (d), FALSE);
+		gchandle = mono_gchandle_new_weakref_from_handle ((MonoObjectHandle)d);
```
Comment (lambdageek) [inline mono/metadata/marshal.c]: `MONO_HANDLE_CAST (MonoObect, d)`

File: mono/metadata/marshal.h [hunk 5]
```diff
@@ -772,11 +772,13 @@ ves_icall_System_ComObject_ReleaseInterfaces(MonoComObject* obj);
 gpointer
 ves_icall_System_ComObject_GetInterfaceInternal (MonoComObject* obj, MonoReflectionType* type, MonoBoolean throw_exception);
 
+TYPED_HANDLE_DECL (MonoComInteropProxy);
```
Comment (lambdageek) [inline mono/metadata/marshal.h]: I try to put the handle declarations near the struct definition try to to encourage folks to notice that it's a managed type with a handle and to write new code using handles, if possible.   In `object-internals.h`.

File: mono/metadata/cominterop.c (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/cominterop.c]: I didn't know of this construct, ok.

File: mono/metadata/marshal.c (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/marshal.c]: oops right

File: mono/metadata/marshal.c (hunk 4 above)
Comment (jaykrell) [reply mono/metadata/marshal.c]: I didn't know of this construct, ok.

File: mono/metadata/marshal.h (hunk 5 above)
Comment (jaykrell) [reply mono/metadata/marshal.h]: I'll try that.

File: mono/metadata/cominterop.c (hunk 2 above)
Comment (jaykrell) [reply mono/metadata/cominterop.c]: I didn't know of this construct, ok.

**Comment by jaykrell** [comment]
@monojenkins build Linux i386

**Comment by jaykrell** [comment]
@monojenkins Linux x64 Interpreter

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Interpreter

**Pull request merged**